src/conoffsetcollect/kafka_offset_collect.go

```diff
@@ -0,0 +1,358 @@
+		}
+	}
+
+	return request
+}
+
+func createFetchRequest(topicPartitions TopicPartitions, client connection.Client) *sarama.FetchRequest {
+	// Explanation:
+	// To add a block, you have to specify an offset to start reading from. Unfortunately, unlike
+	// other parts of the library, this offset cannot use the standard enums OffsetOldest or OffsetNewest.
+	// Thus, in order to make a fetch request, we first have to get the oldest offset with client.GetOffset.
+	// This significantly slows things down, and we will likely need to run this concurrently for decent performance
+
+	request := &sarama.FetchRequest{
+		MaxWaitTime: int32(1000),
+		MinBytes:    int32(0),
+		MaxBytes:    int32(10000),
+		Version:     int16(0),
+		Isolation:   sarama.ReadUncommitted,
+	}
+
+	for topic, partitions := range topicPartitions {
+		for _, partition := range partitions {
+			offset, err := client.GetOffset(topic, partition, sarama.OffsetOldest)
+			if err != nil {
+				log.Info("Failed to get offset for partition")
+			}
+			request.AddBlock(topic, partition, offset, 10000)
+		}
+	}
+
+	return request
+}
+
+// getAllConsumerGroupsFromKafka gets a list of all consumer groups, and populates
+// it with a list of all topics belonging to those consumer groups.
+func getAllConsumerGroupsFromKafka(client connection.Client) (args.ConsumerGroups, error) {
+
+	// Create a broker with an API v9.0 connection
+	broker := client.Brokers()[0]
+	conf := sarama.NewConfig()
+	conf.Version = sarama.V0_9_0_0
+	if err := broker.Close(); err != nil {
+		return nil, err
+	}
+	if err := broker.Open(conf); err != nil {
+		return nil, err
+	}
+
+	// Get a list of all groups
+	groupListResp, err := broker.ListGroups(&sarama.ListGroupsRequest{})
+	if err != nil {
+		return nil, err
+	}
+
+	// Add each group to the request
+	groupDescribeReq := sarama.DescribeGroupsRequest{}
+	for group, t := range groupListResp.Groups {
+		if t == "consumer" {
+			groupDescribeReq.AddGroup(group)
+		}
+	}
+	groupDescribeResp, err := broker.DescribeGroups(&groupDescribeReq)
+	if err != nil {
+		return nil, err
+	}
+
+	// For each group, get its topics
+	consumerGroups := make(args.ConsumerGroups)
+	for _, groupDescription := range groupDescribeResp.Groups {
+		consumerGroups[groupDescription.GroupId] = make(map[string][]int32)
+		for _, memberDescription := range groupDescription.Members {
+			metadata, err := memberDescription.GetMemberMetadata()
+			if err != nil {
+				continue
+			}
+			for _, topic := range metadata.Topics {
+				consumerGroups[groupDescription.GroupId][topic] = make([]int32, 0)
+			}
+		}
+	}
+
+	return consumerGroups, nil
+
+}
+
+// populateOffsetStructs takes a map of offsets and high water marks and
+// populates an array of partitionOffsets which can then be marshalled into metric
+// sets
+func populateOffsetStructs(offsets, hwms groupOffsets) []*partitionOffsets {
+
+	var poffsets []*partitionOffsets
+	for topic, partitions := range offsets {
```

Comment:
I think we want to iterate over `hwms` rather than `offsets` as a High water mark can exist without and offset.